library/lib.ml

```diff
@@ -423,38 +423,40 @@ type section_data = {
   sec_entry : secentry list;
   sec_workl : Opaqueproof.work_list;
   sec_abstr : abstr_list;
+  sec_poly  : bool;
 }
 
-let empty_section_data = {
+let empty_section_data ~poly = {
   sec_entry = [];
   sec_workl = (Names.Cmap.empty,Names.Mindmap.empty);
   sec_abstr = (Names.Cmap.empty,Names.Mindmap.empty);
+  sec_poly  = poly;
 }
 
 let sectab =
   Summary.ref ([] : section_data list) ~name:"section-context"
 
-let add_section () =
-  sectab := empty_section_data :: !sectab
-
-let check_same_poly p vars =
-  let pred = function Context _ -> p = false | Variable {poly} -> p != poly in
-  if List.exists pred vars then
+let check_same_poly p sec =
+  if p != sec.sec_poly then
     user_err Pp.(str  "Cannot mix universe polymorphic and monomorphic declarations in sections.")
 
+let add_section ~poly () =
+  List.iter (fun tab -> check_same_poly poly tab) !sectab;
+  sectab := empty_section_data ~poly :: !sectab
+
 let add_section_variable ~name ~kind ~poly univs =
   match !sectab with
     | [] -> () (* because (Co-)Fixpoint temporarily uses local vars *)
     | s :: sl ->
-       List.iter (fun tab -> check_same_poly poly tab.sec_entry) !sectab;
+       List.iter (fun tab -> check_same_poly poly tab) !sectab;
        let s = { s with sec_entry = Variable {id=name;kind;poly;univs} :: s.sec_entry } in
```

Comment:
The `poly` in `Variable` is now redundant.

Reply:
Will remove.